### PR TITLE
Add scheduled visibility for assessments

### PR DESCRIPTION
**Pull request by lxyhan**
## Proposed Changes

Added datetime-based visibility scheduling for assessments. Instructors can now set `visible_on` and `visible_until` dates to automatically show/hide assessments instead of manually toggling them.

Right now instructors can only mark assessments as visible or hidden with a boolean. They've been requesting the ability to schedule when assessments become visible (like "October 1 - December 31") so they can set it up at the start of the semester and not worry about manually publishing assignments later.

### My changes in this PR

**Database:**
- Added `visible_on` and `visible_until` datetime columns to `assessments` and `assessment_section_properties` tables
- Updated the `check_repo_permissions` SQL function to enforce datetime visibility for git access

**Models:**
- Updated `Student#visible_assessments` to check datetime ranges
- Added validations to ensure `visible_on < visible_until`
- Datetime columns override `is_hidden` when they're set

**API:**
- Exposed `visible_on` and `visible_until` fields in assignments and grade entry forms API endpoints

**Tests:**
- Added validation tests for datetime columns
- Added 18 test cases for visibility logic (global, section-specific, datetime ranges, edge cases)
- Added 9 policy tests for datetime visibility authorization

- If `visible_on`/`visible_until` are set, they override the `is_hidden` boolean
- Section-specific datetime settings override global settings (same pattern as section due dates)
- Backwards compatible - if datetime columns are null, falls back to `is_hidden`
- Can set just `visible_on`, just `visible_until`, or both

## Type of Change
*(Write an `X` or a brief description next to the type or types that best describe your changes.)*

| Type                                                                                    | Applies? |
|-----------------------------------------------------------------------------------------|----------|
| 🚨 *Breaking change* (fix or feature that would cause existing functionality to change) |          |
| ✨ *New feature* (non-breaking change that adds functionality)                          |     x     |
| 🐛 *Bug fix* (non-breaking change that fixes an issue)                                  |          |
| 🎨 *User interface change* (change to user interface; provide screenshots)              |          |
| ♻️ *Refactoring* (internal change to codebase, without changing functionality)          |          |
| 🚦 *Test update* (change that *only* adds or modifies tests)                            |          |
| 📦 *Dependency update* (change that updates a dependency)                               |          |
| 🔧 *Internal* (change that *only* affects developers or continuous integration)         |          |

## Checklist
*(Complete each of the following items for your pull request. Indicate that you have completed an item by changing the `[ ]` into a `[x]` in the raw text, or by clicking on the checkbox in the rendered description on GitHub.)*

Before opening your pull request:

- [x] I have performed a self-review of my changes.
  - Check that all changed files included in this pull request are intentional changes.
  - Check that all changes are relevant to the purpose of this pull request, as described above.
- [x] I have added tests for my changes, if applicable.
  - This is **required** for all bug fixes and new features.
- [x] I have updated the project documentation, if applicable.
  - This is **required** for new features.
- [x] If this is my first contribution, I have added myself to the [list of contributors](https://github.com/MarkUsProject/Markus/blob/master/doc/markus-contributors.txt).

After opening your pull request:

- [x] I have updated the project Changelog (this is required for all changes).
- [x] I have verified that the pre-commit.ci checks have passed.
- [x] I have verified that the CI tests have passed.
- [x] I have reviewed the test coverage changes reported by Coveralls.
- [x] I have [requested a review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) from a project maintainer.

## Questions and Comments
*(Include any questions or comments you have regarding your changes.)*

### CODE REVIEW

File: Changelog.md [hunk 1]
```diff
@@ -5,6 +5,7 @@
 ### 🚨 Breaking changes
 
 ### ✨ New features and improvements
+- Added datetime-based visibility scheduling for assessments with `visible_on` and `visible_until` columns
```
Comment (david-yz-liu) [inline Changelog.md]: Make sure to include the PR number here

File: app/models/assessment.rb [hunk 2]
```diff
@@ -38,6 +41,14 @@ def short_identifier_unchanged
     false
   end
 
+  def visible_dates_are_valid
+    return if visible_on.nil? || visible_until.nil?
+    return unless visible_on.respond_to?(:>=) && visible_until.respond_to?(:>=)
```
Comment (david-yz-liu) [inline app/models/assessment.rb]: I don't think this line is necessary, if it's reach then shouldn't these attributes be dates?

File: app/models/assessment_section_properties.rb [hunk 3]
```diff
@@ -5,6 +5,10 @@ class AssessmentSectionProperties < ApplicationRecord
 
   has_one :course, through: :assessment
   validate :courses_should_match
+  validate :visible_dates_are_valid
```
Comment (david-yz-liu) [inline app/models/assessment_section_properties.rb]: Validations are typically run in top-down order, so it would be more typical to write this validation below the next two (so that the types are checked first)

File: app/models/student.rb [hunk 4]
```diff
@@ -256,26 +256,70 @@ def grace_credits_used_for(assessment)
...
-      visible = visible.where(is_hidden: false)
+      # No section assigned - just check global visibility
+      visible = visible.where(
+        '(assessments.visible_on IS NULL OR assessments.visible_on <= ?) AND ' \
+        '(assessments.visible_until IS NULL OR assessments.visible_until >= ?)',
+        current_time, current_time
+      )
+
+      # Use datetime if set, otherwise fall back to is_hidden
+      visible.where(
```
Comment (david-yz-liu) [inline app/models/student.rb]: This can be simplified using built-in ActiveRecord methods. In this case you can use both `.where.not(...)` and `.or`

File: app/models/student.rb [hunk 5]
```diff
@@ -256,26 +256,70 @@ def grace_credits_used_for(assessment)
...
+        '(assessments.visible_on IS NOT NULL OR assessments.visible_until IS NOT NULL) OR ' \
+        'assessments.is_hidden = false'
+      )
+
+      section_visible.or(global_visible)
     else
-      visible = visible.where(is_hidden: false)
+      # No section assigned - just check global visibility
+      visible = visible.where(
+        '(assessments.visible_on IS NULL OR assessments.visible_on <= ?) AND ' \
```
Comment (david-yz-liu) [inline app/models/student.rb]: Similar to my comments below, I'd prefer using the ActiveRecord methods where possible. In this case the part with the `<=` is fine, but you can use `.or` instead of SQL `OR` and chained `.where`s instead of `AND`

File: app/models/student.rb [hunk 6]
```diff
@@ -256,26 +256,72 @@ def grace_credits_used_for(assessment)
...
+                        .or(section_visible.where(assessment_section_properties: { visible_on: ..current_time }))
+      section_visible = section_visible
+                        .where(assessment_section_properties: { visible_until: nil })
+                        .or(section_visible.where(assessment_section_properties: { visible_until: current_time.. }))
+
+      # Use datetime if set, otherwise fall back to is_hidden
+      section_visible = section_visible
+                        .where.not(assessment_section_properties: { visible_on: nil })
+                        .or(section_visible.where.not(assessment_section_properties: { visible_until: nil }))
+                        .or(section_visible.where(assessment_section_properties: { is_hidden: false }))
```
Comment (david-yz-liu) [inline app/models/student.rb:295]: This is something subtle here with the `is_hidden` section property.

In the case that for the section properties, `visible_on: nil, visible_until: nil, is_hidden: true`, then the assignment should _not_ be visible to students in that section, regardless of what the global properties are. I don't think this is currently being implemented correctly by this logic.

You handled this differently in the `visibility_hash` method by first checking whether any of the three section properties were non-nil before overriding the global properties.

And in the SQL function, you're only checking `assessment_section_properties.is_hidden IS NOT NULL` but should probably be checking all three.

File: spec/models/student_spec.rb [hunk 7]
```diff
@@ -395,4 +395,113 @@
...
+        assignment = create(:assignment, course: course, is_hidden: false)
+        expect(student.visible_assessments).to include(assignment)
+      end
+
+      it 'hides assessment when is_hidden=true and no datetime set' do
+        assignment = create(:assignment, course: course, is_hidden: true)
+        expect(student.visible_assessments).not_to include(assignment)
+      end
+
+      it 'shows assessment when datetime range is valid (overrides is_hidden=true)' do
```
Comment (david-yz-liu) [inline spec/models/student_spec.rb]: The word "valid" is not useful in this context. Better wording would be "when the datetime range includes the current time". This repeats a few times in the test descriptions.

File: app/models/student.rb [hunk 8]
```diff
@@ -256,26 +256,78 @@ def grace_credits_used_for(assessment)
...
     if self.section_id
       visible = visible.left_outer_joins(:assessment_section_properties)
                        .where('assessment_section_properties.section_id': [self.section_id, nil])
-      visible = visible.where('assessment_section_properties.is_hidden': false)
-                       .or(visible.where('assessment_section_properties.is_hidden': nil,
-                                         'assessments.is_hidden': false))
+
+      # Check section-specific visibility first (takes precedence when any section property is set)
+      section_visible = visible.where('assessment_section_properties.section_id': self.section_id)
+                               .where('assessment_section_properties.is_hidden IS NOT NULL OR ' \
```
Comment (david-yz-liu) [inline app/models/student.rb]: this is a place where you can use ActiveRecord methods like `.or` and `.not`

File: app/models/student.rb [hunk 9]
```diff
@@ -256,26 +256,78 @@ def grace_credits_used_for(assessment)
...
+
+      # Use datetime if set, otherwise fall back to is_hidden
+      section_visible = section_visible
+                        .where.not(assessment_section_properties: { visible_on: nil })
+                        .or(section_visible.where.not(assessment_section_properties: { visible_until: nil }))
+                        .or(section_visible.where(assessment_section_properties: { is_hidden: false }))
+
+      # Check global visibility (when no section-specific settings exist)
+      global_visible = visible.where('assessment_section_properties.section_id': nil)
+                              .or(visible.where('assessment_section_properties.is_hidden IS NULL AND ' \
```
Comment (david-yz-liu) [inline app/models/student.rb]: a similar comment here, you can both use `: nil` and pass multiple conditions to `.where` instead of relying on raw SQL

**Review by david-yz-liu — COMMENTED** [review]
@lxyhan nice work, I just left two small comments
